### PR TITLE
Votebuttons

### PR DESCRIPTION
**Pull request by omaribrahim4**
Added Upvote/Downvote Functionality

### CODE REVIEW

File: src/main/java/com/AjaxController.java [hunk 1]
```diff
@@ -1,107 +1,158 @@
...
+
+    /*
+     * Add perk to a users subscription.
+     * @Param: userName=userName of user
+     * @Param: subName=name of subscription
+     * @Param: perkJson=Json of the perk to be added
+     * */
+    @ResponseStatus(HttpStatus.OK)
+    @RequestMapping(method=RequestMethod.POST, value = "/AddPerk",consumes = {"application/json"})
+    public void AddPerk(@RequestParam String userName,@RequestParam String subName,@RequestBody Perk perkJson) {
```
Comment (DanHoganRed) [inline src/main/java/com/AjaxController.java:91]: It looks like you have an old AddPerk method. Can you change it back?

File: src/main/java/com/AjaxController.java [hunk 2]
```diff
@@ -1,107 +1,158 @@
...
+    private UserService userService;
+    @Autowired
+    private PerkService perkService;
+    @Autowired
+    private SubscriptionService subscriptionService;
+
+    User user;
+
+    /**
+    public static final Comparator<Perk> DESCENDING_COMPARATOR = new Comparator<Perk>() {
```
Comment (DanHoganRed) [inline src/main/java/com/AjaxController.java:26]: If this is dead code please remove.

File: src/main/resources/static/MainPage.js [hunk 3]
```diff
@@ -1,155 +1,223 @@
...
+                dialog_sub.dialog( "close" );
+            }
+            return valid;
+        }
+
+        /* Add perk for the subscription. */
+    function addPerk() {
+        var valid = true;
+
+        valid = valid && checkLength( name_perk, 3 );
```
Comment (DanHoganRed) [inline src/main/resources/static/MainPage.js]: Can you fix the spacing please?

File: src/main/java/com/PerkController.java [hunk 4]
```diff
@@ -1,14 +1,35 @@
...
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.*;
 
 @Controller
 public class PerkController {
 
+    @Autowired
     private PerkService perkService;
+
```
Comment (DanHoganRed) [inline src/main/java/com/PerkController.java]: If you are going to use the PerkController like this it needs to be a RestController.

File: src/main/resources/static/MainPage.js [hunk 5]
```diff
@@ -1,155 +1,223 @@
...
+            var subname = $(event.target).parent().parent().find("#subscription_name").text();
+            var perkname = $(event.target).parent().parent().find("#perk_name").text();
+
+            /* Construct JSON and send. This call is not async because the calls to GetTable finishes before this one.  */
+            var perkJson = {"name":perkname,"description":desc_perk.val(),"expiryDate":null,"subscription":null};
+            $.ajax({
+                type:"POST",
+                async:false,
+                contentType: "application/json; charset=utf-8",
+                url: cntxPath+"/upvote?userName="+userName+"&subName="+subname,
```
Comment (DanHoganRed) [inline src/main/resources/static/MainPage.js:116]: You don't use either the userName or subName parameters in the controller.

File: src/main/resources/static/MainPage.js [hunk 6]
```diff
@@ -1,155 +1,223 @@
...
+            var subname = $(event.target).parent().parent().find("#subscription_name").text();
+            var perkname = $(event.target).parent().parent().find("#perk_name").text();
+
+            /* Construct JSON and send. This call is not async because the calls to GetTable finishes before this one.  */
+            var perkJson = {"name":perkname,"description":desc_perk.val(),"expiryDate":null,"subscription":null};
+            $.ajax({
+                type:"POST",
+                async:false,
+                contentType: "application/json; charset=utf-8",
+                url: cntxPath+"/downvote?userName="+userName+"&subName="+subname,
```
Comment (DanHoganRed) [inline src/main/resources/static/MainPage.js:140]: You don't use either the userName or subName parameters in the controller.

File: src/main/resources/static/MainPage.js [hunk 7]
```diff
@@ -1,155 +1,223 @@
...
+                dataType:"json",
+                data: JSON.stringify(perkJson)
+            });
+            /* Update table. */
+            GetTable();
+        });
+    }
+
+    /**
+    function addDownvoteListener(){
```
Comment (DanHoganRed) [inline src/main/resources/static/MainPage.js]: Remove dead code

**Review by DanHoganRed — CHANGES_REQUESTED** [review]
Please see my comments
Additionally you need to write tests if you change anything in the controllers. 
There are also conflicts with the merge but I can take care of that later.
For future reference make sure you pull before writing code and try not to make changes to the entire file so it's easier to see changes for the review.

File: src/main/java/com/AjaxController.java (hunk 2 above)
Comment (omaribrahim4) [reply src/main/java/com/AjaxController.java:26]: I kept this to use it later for sorting the perks by score.

File: src/main/resources/static/MainPage.js (hunk 6 above)
Comment (omaribrahim4) [reply src/main/resources/static/MainPage.js:140]: Done. Applied changes to PerkController.

File: src/main/resources/static/MainPage.js (hunk 5 above)
Comment (omaribrahim4) [reply src/main/resources/static/MainPage.js:116]: Done. Applied changes to PerkController

File: src/main/java/com/AjaxController.java (hunk 1 above)
Comment (omaribrahim4) [reply src/main/java/com/AjaxController.java:91]: Done.

File: src/main/java/com/PerkController.java (hunk 4 above)
Comment (omaribrahim4) [reply src/main/java/com/PerkController.java]: Done.

File: src/main/resources/static/MainPage.js (hunk 3 above)
Comment (omaribrahim4) [reply src/main/resources/static/MainPage.js]: Done.

File: src/main/resources/static/MainPage.js (hunk 7 above)
Comment (omaribrahim4) [reply src/main/resources/static/MainPage.js]: Done.

**Review by DanHoganRed — APPROVED** [review]

**Comment by DanHoganRed** [comment]
Looks good I made issue #48 for you to address the tests.

**Pull request merged**